### DISCUSSION TITLE
[Validator] How to make Assert\Sequentially stop after first violation when using multiple validation groups

### DISCUSSION THREAD

**Discussion by annuh** [symfony/symfony#49934] [Q&A] [answered]
Hi!

I would like to validate a given email address via an external email validator during registration, when the email is first validated (and accepted) by other email validators.

For this, I probably(?) need to combine Assert\Sequentially with an extra validation groups (`registration`). However, the validation results in different then I expected.

Here is the code (simplified) I'm using:

```php
# User.php
use Symfony\Component\Validator\Constraints as Assert;

    #[Assert\Sequentially([
        new Assert\NotBlank(),
        new Assert\Email(mode: 'html5'),
        new Assert\Email(mode: 'strict'),
        new App\Validator\Constraints\Email\ExternalCheck(groups: ['registration']),
    ])]
    private string $email;
```

```php
# RegistrationController.php
use Symfony\Component\Validator\Constraint;
use Symfony\Component\Validator\Validator\ValidatorInterface;

public function test(ValidatorInterface $validator)
{
    $user = new User();
    $user->setEmail('');
    
    $errors = $this->validator->validate($user, null, [Constraint::DEFAULT_GROUP, 'registration']);
}
```

I expected `$errors` would contain only one violation: the `NotBlank`, but instead it contains two errors: the `NotBlank` and `ExternalCheck`. A work-around for this is to apply the validator for each validation group separate:
```php
$errors = $this->validator->validate($user, null, [Constraint::DEFAULT_GROUP]);
if ($errors->count() === 0) {
    $errors = $this->validator->validate($user, null, ['registration']);
}
```

Is this actual expected behavior? If so, what is the recommended way to run multiple validators sequentially with different validation groups?

Thanks!

**Comment by elevado** [comment]
Hi, 

I think [Sequentially](https://symfony.com/doc/current/reference/constraints/Sequentially.html) is what you are looking for.

**Reply by annuh** [reply]
Thanks for your reply. Sequentially is what I've used indeed (see the code above), but when using different validation groups (`'Default' + 'registration'`) the validation result seems unexpected.

**Reply by elevado** [reply]
Oops, sorry 😕 ! That's right, I had Group in mind.  I'll have a look at it. Maybe I can help.

**Reply by elevado** [reply]
@annuh Maybe I have misunderstood something. Sequentially works as expected, with and without groups. The validator breaks on the first violation.

```php
use PHPUnit\Framework\TestCase;
use Symfony\Component\Validator\Constraints\Length;
use Symfony\Component\Validator\Constraints\NotBlank;
use Symfony\Component\Validator\Constraints\Sequentially;
use Symfony\Component\Validator\ValidatorBuilder;

$validator = (new ValidatorBuilder())->getValidator();
$violations = $validator->validate(
    'aaaaaa', // 6 chars
    [
        new Sequentially(
            [
                new Length(min: 4, minMessage: 'min 4'),
                new Length(min: 6, minMessage: 'min 6'),
                new Length(min: 8, minMessage: 'min 8'), // <---- violation
                new NotBlank(message: 'not blank'),
                new Length(min: 10, minMessage: 'min 10', groups: ['registration'])
            ]
        )
    ]
);

$countErrors = $violations->count();
// $countErrors -> 1
foreach ($violations as $violation) {
    $errorMessage = $violation->getMessage();
}

// $errorMessage -> 'min 8';
```

**Reply by stof** [reply]
@elevado your code validates only the default group

**Reply by elevado** [reply]
@stof Oh yes, right. Now I understand what the challenge is.

@annuh Now I remembered that I had a similar problem once.  I implemented my custom validator where I used the validator to validate with the constraints. First I validated the value and if no error, then I used the validator with the constraint to check the value via Api.

**Comment by ChubongaGithub** [comment, marked as answer]
I believe what you want is [GroupSequence](https://symfony.com/doc/current/validation/sequence_provider.html).

Can be used like this:
`$errors = $this->validator->validate($user, null, new GroupSequence(['Default', 'registration']));`

**Reply by annuh** [reply]
Using a `GroupSequence` solved this problem indeed, thank you!
